skimage/_shared/coord.py

```diff
@@ -59,7 +59,7 @@ def _ensure_spacing(coord, spacing, p_norm, max_out):
 
 
 def ensure_spacing(coords, spacing=1, p_norm=np.inf, min_split_size=50,
-                   max_out=None):
+                   max_out=None, *, max_split_size=2000):
```

Comment:
the fix below looks okay to me. out of curiosity, how did you select this value of 2000? (is it based on keeping things under a particular memory limit?)

Reply:
By rule of thumb :sweat_smile:. I profiled `cKDTree` using
```python
In [9]: coords = np.random.randint(low=0, high=1848, size=(60000, 2))     
In [10]: for n in range(1000, 60001, 1000): 
    ...:     pts = coords[:n, :] 
    ...:     %memit cKDTree(pts).query_ball_point(pts, r=100, p=np.inf) 
    ...:     %timeit cKDTree(pts).query_ball_point(pts, r=100, p=np.inf) 
```
and obtained

![Figure_1](https://user-images.githubusercontent.com/3438227/141964064-5a345e15-1bca-4da3-8038-ae655ae7628d.png)

Maybe 10000 is better?

Reply:
@rfezzani I think this profiling is not sophisticated enough, I'd instead propose the full shebang:

```python
for n in range(1000, 60001, 1000):
    %timeit ensure_spacing(coords, spacing=100, min_split_size=50, max_split_size=n)
    %memit ensure_spacing(coords, spacing=100, min_split_size=50, max_split_size=n)
```

The key is that in our typical situation, we are whittling down the size of the cKDTree as we go along, so we can afford bigger sizes towards the end, as we will be effectively doing `cKDTree(accepted_pts).query(remaining_pts)`, where `accepted_pts` is smaller than `remaining_pts`.

Reply:
Based on the following script and output, the choice of 2000 here seems good to me. Values < 1000 become slower and values > 10k started substantially increasing memory use as well as becoming slower. Perhaps differently distributed coordinates would behave in a different manner, but this seems like a good default.

###Script
```Python
import numpy as np

from skimage._shared.coord import ensure_spacing

coords = np.random.randint(low=0, high=1848, size=(64000, 2))

for n in [250, 500, 1000, 2000, 4000, 8000, 12000, 16000, 32000, 64000]:
    print(f"\nmax_split_size = {n}")
    %timeit ensure_spacing(coords, spacing=100, min_split_size=50, max_split_size=n)
    %memit ensure_spacing(coords, spacing=100, min_split_size=50, max_split_size=n)
```
### Script Output
```
max_split_size = 250
2.24 s ± 144 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
peak memory: 167.72 MiB, increment: 0.11 MiB

max_split_size = 500
1.71 s ± 85.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
peak memory: 167.86 MiB, increment: 0.00 MiB

max_split_size = 1000
1.43 s ± 51.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
peak memory: 168.36 MiB, increment: 0.00 MiB

max_split_size = 2000
1.43 s ± 60.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
peak memory: 169.87 MiB, increment: 0.50 MiB

max_split_size = 4000
1.47 s ± 94.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
peak memory: 176.54 MiB, increment: 5.42 MiB

max_split_size = 8000
1.7 s ± 44.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
peak memory: 204.29 MiB, increment: 23.70 MiB

max_split_size = 12000
1.94 s ± 61.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
peak memory: 253.79 MiB, increment: 73.20 MiB

max_split_size = 16000
2.01 s ± 98.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
peak memory: 313.06 MiB, increment: 95.39 MiB

max_split_size = 32000
2.21 s ± 86.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
peak memory: 552.30 MiB, increment: 238.87 MiB

max_split_size = 64000
4.8 s ± 131 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
peak memory: 2431.14 MiB, increment: 1535.70 MiB
```

Reply:
Wow @rfezzani your thumb found the exact minimum of the function! 😂 👏 👏 👏  Thanks @grlee77 for running this! 🚀 Very valuable info. Perhaps it's worth adding a note to the docstring linking to this comment?

Reply:
Sorry @jni, I can't find time to answer :pray: Thank you @grlee77 for taking over :wink:
I